### ISSUE TITLE
Modernize the CI/CD by migrating travis.yml to .github/workflows/CI.yml

### ISSUE THREAD

**Issue by reactive-firewall** [mbrossard/pkcs11#4]
## Tasks:
 ~- [ ] assign to @reactive-firewall~ (_optional_)
 - [x] readup on Travis CI/CD config
 - [x] re-write for GHA (_see below for details_)
 - [x] Test on feature branch (Caveat: Expected failure without further maintenance to core project codebase 🙈 )
 - [ ] Submit PR/MR review and hand off ticket . . .
 ~- [ ] assign back to @mbrossard~ (_optional_)
 - [ ] Merge back (_optional_)
 
 ---
 
 ### Re-write for GHA
 
 - [x] re-write for GHA
   - [x] Initial re-write for GHA (before debug)
   - [x] Read up on `libsoftokn3`
   <details>
   <summary>Here are some useful (albeit OLD) links</summary>

   **NSS Documentation**: This includes detailed documentation on how to use NSS, including its libraries and APIs.
[NSS Documentation](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/NSS)

   **Source Code Repository**: As we are looking for implementation details, we can also check the source code, which is hosted on GitHub.
[NSS GitHub Repository](https://github.com/nss-dev/nss)

   </details>

   - [x] find and document ~**ALL**~ broken `C_Initialize` definition in project (e.g. `/* C_Initialize calls are broken see mbrossard/pkcs11#4 */`)
   ~- [ ] read up on the project `secmod.db` in order to understand codebase better~
   - [x] re-write GHA environment to include `NSS_LIB_PARAMS="configdir='$PWD' certPrefix='' keyPrefix='' secmod='secmod.db'"`
   - [x] Misc. (e.g. cleanup from complexity fan-out)
   - [x] report back on this GHI summary findings
   
### Migration from libssl3-dev to libssl-dev

  * If you were previously using `libssl3-dev` with OpenSSL 1.1, we should be able to migrate to `libssl-dev` for `OpenSSL` 3.0 without significant issues, but may need to review the code for every deprecated functions or changes in the API between `OpenSSL` 1.1 and `OpenSSL` 3.0.

* As already mentioned: OpenSSL 3.0 introduced some changes, including a new licensing model and some API changes. (clarification from suspected to confirmed)

> [!TIP]
> we can probably use simple conditionals on the incorrect pointer types:
> ```c
>     #if (OPENSSL_VERSION_NUMBER & 0xF0000000) >= 0x30000000
>        // Code for when OPENSSL_VERSION_MAJOR is at least 3
>     #else
>        // Code for when OPENSSL_VERSION_MAJOR is less than 3
>     #endif
> ```
> see [definition](https://github.com/openssl/openssl/blob/master/include/openssl/opensslv.h.in)

### Summary

 * On Ubuntu 22.04, libssl-dev corresponds to OpenSSL 3.0.
 * We can check the specific version using `apt-cache policy libssl-dev`.
 
 * we can ignore the deprecated warnings (at least for the moment) as long as openssl does not define `OPENSSL_NO_DEPRECATED_3_0` they are just deprecated attributes and the functions are still present.
 
 * 🤔 the Travis CI is migrated to GHA (caveat: expected failures due to discussed API changes in dependencies 🙊 🙈 )
 * 👍🏻 main goal accomplished
 * :-1: Not sure I'm comfortable tweaking the more security relevant code with my level of understanding of `cryptokit` and `NSSlib` (or rather lack-there-of)
 * I think I'll clean up the circleci config before anymore attempts at debugging the modernization issues (debugging the main code, if left to me alone, will have significant learning time-lag as I read all the things, so I want a solid test workflow setup in CI first)

**Comment by reactive-firewall** [comment]
@mbrossard,

🤔 ok some things have changed in openssl and gcc 🙊

Of note:
 - GHA CI/CD runners need a gcc version between 10 and 12 (caveat: macOS can find a 13 and 14)
 - openssl v3 (the default now) deprecated a bunch of stuff see: [these logs in CI/CD](https://github.com/reactive-firewall/pkcs11/actions/runs/12564603353/job/35027894500#step:6:223) (the build warnings)
   - if you don't plan to update the code, can you point me in the right direction for me to start learning the relevant API so I can debug this? :bow:
   
ignore the issue tag commit list, I'm just using `--amend` while I debug.
~The working GHA with the failing build/tests? are on the git commit:
2f08d3a21e5c132ed11f8e9d43288214895d7417~

:bow: Hope this helps!

**Comment by mbrossard** [comment]
It seems `libsoftokn3` has changed behavior. The current method that relied on fallbacks from C_Initialize calls do not work anymore. If you want (some) tests to succeed you need to set this environment variable: `NSS_LIB_PARAMS="configdir='$PWD' certPrefix='' keyPrefix='' secmod='secmod.db'"`

There are also some other changes to OpenSSL (the command-line arguments to `genrsa` and the ECDSA-support detection). It's probably easier to assume OpenSSL version 3.x.

**Comment by reactive-firewall** [comment]
@mbrossard,

In my spare time, I've made a little bit of progress on this, and wanted to make a note of the status on this:

I have the ubuntu builds completing (with lots of warnings to review and cleanup) but the tests are doomed to fail at the moment because the module is not being configured correctly for testing,

The symptom I'm seeing at the moment is lack of the module being defined.

<details><summary>TL;DR please clarify the intent of the Make Variable `DEFAULT_PKCS11_MODULE`</summary>

I only seem to get a blank value for the Make Variable `DEFAULT_PKCS11_MODULE` from the generated makefile here:

https://github.com/reactive-firewall/pkcs11/blob/7ea3daef309399e8442d3611d172f14d0b2ed857/Makefile.am#L109-L109

and I suspect this should line up with the code in or around:

https://github.com/reactive-firewall/pkcs11/blob/7ea3daef309399e8442d3611d172f14d0b2ed857/src/cryptoki.c#L90-L92

Rationale: this would make sense for identifying the module at test time where the CI run fails for now. See screenshots:

<img width="640" alt="Image" src="https://github.com/user-attachments/assets/0757363c-346d-4b4b-9a2b-8f44cb88f374" />

**AND**

<img width="797" alt="Image" src="https://github.com/user-attachments/assets/f8dfc72f-3acc-4b7c-b223-8b4569f4933b" />

I'm more comfortable with BSD make and typically don't use `autoconf` or `automake` myself, so I'm
not sure what was intended by the config that I believe should be setting the value:

https://github.com/reactive-firewall/pkcs11/blob/7ea3daef309399e8442d3611d172f14d0b2ed857/configure.ac#L66-L103

My guess it that it should accept some argument (not sure where?) first, and failing that is should perform the best-effort search (which is the part I can follow)

The same references in your project can be found with this:
https://github.com/search?q=repo%3Ambrossard%2Fpkcs11%20DEFAULT_PKCS11_MODULE&type=code

Thanks for reading this far :bow: If you have some time, please clarify the intent of the Make Variable `DEFAULT_PKCS11_MODULE` for me.

</details>
